stacktracer/jvm/src/main/scala/scalaz/zio/internal/stacktracer/impl/AkkaLineNumbers.scala

```diff
@@ -0,0 +1,374 @@
+/*
+ * Copyright (C) 2014-2019 Lightbend Inc. <https://www.lightbend.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package scalaz.zio.internal.stacktracer.impl
+
+import java.io.{DataInputStream, InputStream}
+import java.lang.invoke.SerializedLambda
+
+import scala.annotation.switch
+import scala.util.control.NonFatal
+
+/**
+ * This parser has been copied from the `akka-typed` project and modified,
+ * as always these guys are way ahead of us! ;)
+ *
+ * https://github.com/akka/akka/blob/4729a80e7e6f2923a7e6369b21b2693f82c64589/akka-actor/src/main/scala/akka/util/LineNumbers.scala
+ *
+ * ==
+ * This is a minimized byte-code parser that concentrates exclusively on line
+ * numbers and source file extraction. It works for all normal classes up to
+ * format 52:0 (JDK8), and it also works for Lambdas that are Serializable. The
+ * latter restriction is due to the fact that the proxy object generated by
+ * LambdaMetafactory otherwise contains no information about which method backs
+ * this particular lambda (and there might be multiple defined within a single
+ * class).
+ **/
+object AkkaLineNumbers {
```

Comment:
Could you please provide tracer benchmark numbers in a PR comment? It's not clear right away what are the benefits of using this code instead of ASM based parser (outside of maintaining a zero dep status).

Reply:
Asm is a bih faster but this parser has no external dependencies. And this is the primary reason to choose this.

Also keep in mind that once the cache is filled this code will never be invoked.

Reply:
@pshirshov The cache is unbounded and could be a potential problem on large codebases. Do you have plans to optimize it? One of optimization from top of my head is to use String.intern() method for file, clazz maybe

Reply:
Check the whole patch, interning is there already.

Reply:
Didn't notice it, my bad. Cool!

Reply:
@artempyanykh Benchmark numbers on my machine:
```
[info] Running (fork) org.openjdk.jmh.Main scalaz.zio.stacktracer.TracersBenchmark -i 10 -wi 10 -f1 -t8
[info] # JMH version: 1.21
[info] # VM version: JDK 1.8.0_181, Java HotSpot(TM) 64-Bit Server VM, 25.181-b13
[info] # VM invoker: /Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/bin/java
[info] # VM options: <none>
[info] # Warmup: 10 iterations, 1 s each
[info] # Measurement: 10 iterations, 1 s each
[info] # Timeout: 10 min per iteration
[info] # Threads: 8 threads, will synchronize iterations
[info] # Benchmark mode: Average time, time/op
[info] # Benchmark: scalaz.zio.stacktracer.TracersBenchmark.akkaLineNumbers1000Lambdas
[info] # Run progress: 0.00% complete, ETA 00:01:00
[info] # Fork: 1 of 1
[info] # Warmup Iteration   1: 688.078 ?(99.9%) 333.396 ms/op
[info] # Warmup Iteration   2: 539.822 ?(99.9%) 10.256 ms/op
[info] # Warmup Iteration   3: 526.389 ?(99.9%) 10.554 ms/op
[info] # Warmup Iteration   4: 477.833 ?(99.9%) 17.939 ms/op
[info] # Warmup Iteration   5: 481.406 ?(99.9%) 16.112 ms/op
[info] # Warmup Iteration   6: 480.685 ?(99.9%) 19.045 ms/op
[info] # Warmup Iteration   7: 493.985 ?(99.9%) 22.763 ms/op
[info] # Warmup Iteration   8: 490.212 ?(99.9%) 15.283 ms/op
[info] # Warmup Iteration   9: 502.920 ?(99.9%) 29.178 ms/op
[info] # Warmup Iteration  10: 497.979 ?(99.9%) 25.670 ms/op
[info] Iteration   1: 506.635 ?(99.9%) 26.874 ms/op
[info] Iteration   2: 511.377 ?(99.9%) 34.289 ms/op
[info] Iteration   3: 519.323 ?(99.9%) 42.639 ms/op
[info] Iteration   4: 499.538 ?(99.9%) 29.123 ms/op
[info] Iteration   5: 517.134 ?(99.9%) 22.145 ms/op
[info] Iteration   6: 498.225 ?(99.9%) 17.942 ms/op
[info] Iteration   7: 526.698 ?(99.9%) 15.620 ms/op
[info] Iteration   8: 502.834 ?(99.9%) 18.593 ms/op
[info] Iteration   9: 522.352 ?(99.9%) 23.818 ms/op
[info] Iteration  10: 490.843 ?(99.9%) 16.605 ms/op
[info] Result "scalaz.zio.stacktracer.TracersBenchmark.akkaLineNumbers1000Lambdas":
[info]   509.496 ?(99.9%) 17.791 ms/op [Average]
[info]   (min, avg, max) = (490.843, 509.496, 526.698), stdev = 11.768
[info]   CI (99.9%): [491.705, 527.287] (assumes normal distribution)
[info] # JMH version: 1.21
[info] # VM version: JDK 1.8.0_181, Java HotSpot(TM) 64-Bit Server VM, 25.181-b13
[info] # VM invoker: /Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/bin/java
[info] # VM options: <none>
[info] # Warmup: 10 iterations, 1 s each
[info] # Measurement: 10 iterations, 1 s each
[info] # Timeout: 10 min per iteration
[info] # Threads: 8 threads, will synchronize iterations
[info] # Benchmark mode: Average time, time/op
[info] # Benchmark: scalaz.zio.stacktracer.TracersBenchmark.akkaTracer1000Lambdas
[info] # Run progress: 33.33% complete, ETA 00:01:11
[info] # Fork: 1 of 1
[info] # Warmup Iteration   1: 810.213 ?(99.9%) 344.045 ms/op
[info] # Warmup Iteration   2: 610.676 ?(99.9%) 32.374 ms/op
[info] # Warmup Iteration   3: 557.663 ?(99.9%) 30.693 ms/op
[info] # Warmup Iteration   4: 583.463 ?(99.9%) 24.476 ms/op
[info] # Warmup Iteration   5: 526.098 ?(99.9%) 19.209 ms/op
[info] # Warmup Iteration   6: 553.463 ?(99.9%) 12.259 ms/op
[info] # Warmup Iteration   7: 539.786 ?(99.9%) 22.442 ms/op
[info] # Warmup Iteration   8: 553.602 ?(99.9%) 21.540 ms/op
[info] # Warmup Iteration   9: 535.725 ?(99.9%) 26.722 ms/op
[info] # Warmup Iteration  10: 556.537 ?(99.9%) 18.868 ms/op
[info] Iteration   1: 536.602 ?(99.9%) 12.564 ms/op
[info] Iteration   2: 558.272 ?(99.9%) 16.455 ms/op
[info] Iteration   3: 525.707 ?(99.9%) 19.731 ms/op
[info] Iteration   4: 559.580 ?(99.9%) 29.885 ms/op
[info] Iteration   5: 532.769 ?(99.9%) 20.721 ms/op
[info] Iteration   6: 556.363 ?(99.9%) 8.802 ms/op
[info] Iteration   7: 565.317 ?(99.9%) 15.784 ms/op
[info] Iteration   8: 576.543 ?(99.9%) 9.204 ms/op
[info] Iteration   9: 548.832 ?(99.9%) 24.991 ms/op
[info] Iteration  10: 578.714 ?(99.9%) 13.820 ms/op
[info] Result "scalaz.zio.stacktracer.TracersBenchmark.akkaTracer1000Lambdas":
[info]   553.870 ?(99.9%) 27.057 ms/op [Average]
[info]   (min, avg, max) = (525.707, 553.870, 578.714), stdev = 17.897
[info]   CI (99.9%): [526.813, 580.927] (assumes normal distribution)
[info] # JMH version: 1.21
[info] # VM version: JDK 1.8.0_181, Java HotSpot(TM) 64-Bit Server VM, 25.181-b13
[info] # VM invoker: /Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/bin/java
[info] # VM options: <none>
[info] # Warmup: 10 iterations, 1 s each
[info] # Measurement: 10 iterations, 1 s each
[info] # Timeout: 10 min per iteration
[info] # Threads: 8 threads, will synchronize iterations
[info] # Benchmark mode: Average time, time/op
[info] # Benchmark: scalaz.zio.stacktracer.TracersBenchmark.asmTracer1000Lambdas
[info] # Run progress: 66.67% complete, ETA 00:00:34
[info] # Fork: 1 of 1
[info] # Warmup Iteration   1: 392.895 ?(99.9%) 175.151 ms/op
[info] # Warmup Iteration   2: 335.377 ?(99.9%) 53.985 ms/op
[info] # Warmup Iteration   3: 673.830 ?(99.9%) 25.490 ms/op
[info] # Warmup Iteration   4: 877.822 ?(99.9%) 194.560 ms/op
[info] # Warmup Iteration   5: 329.584 ?(99.9%) 16.764 ms/op
[info] # Warmup Iteration   6: 229.133 ?(99.9%) 10.911 ms/op
[info] # Warmup Iteration   7: 249.190 ?(99.9%) 13.161 ms/op
[info] # Warmup Iteration   8: 224.898 ?(99.9%) 55.644 ms/op
[info] # Warmup Iteration   9: 279.752 ?(99.9%) 15.074 ms/op
[info] # Warmup Iteration  10: 207.423 ?(99.9%) 5.639 ms/op
[info] Iteration   1: 240.831 ?(99.9%) 12.544 ms/op
[info] Iteration   2: 228.104 ?(99.9%) 13.935 ms/op
[info] Iteration   3: 181.641 ?(99.9%) 4.350 ms/op
[info] Iteration   4: 194.082 ?(99.9%) 14.451 ms/op
[info] Iteration   5: 262.669 ?(99.9%) 13.056 ms/op
[info] Iteration   6: 241.883 ?(99.9%) 9.852 ms/op
[info] Iteration   7: 210.150 ?(99.9%) 3.828 ms/op
[info] Iteration   8: 359.880 ?(99.9%) 33.290 ms/op
[info] Iteration   9: 209.194 ?(99.9%) 5.922 ms/op
[info] Iteration  10: 260.227 ?(99.9%) 9.506 ms/op
[info] Result "scalaz.zio.stacktracer.TracersBenchmark.asmTracer1000Lambdas":
[info]   238.866 ?(99.9%) 76.064 ms/op [Average]
[info]   (min, avg, max) = (181.641, 238.866, 359.880), stdev = 50.312
[info]   CI (99.9%): [162.802, 314.930] (assumes normal distribution)
[info] # Run complete. Total time: 00:01:38
[info] REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
[info] why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
[info] experiments, perform baseline and negative tests that provide experimental control, make sure
[info] the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
[info] Do not assume the numbers tell you what you want them to tell.
[info] Benchmark                                    Mode  Cnt    Score    Error  Units
[info] TracersBenchmark.akkaLineNumbers1000Lambdas  avgt   10  509.496 ? 17.791  ms/op
[info] TracersBenchmark.akkaTracer1000Lambdas       avgt   10  553.870 ? 27.057  ms/op
[info] TracersBenchmark.asmTracer1000Lambdas        avgt   10  238.866 ? 76.064  ms/op
[success] Total time: 112 s, completed 22-May-2019 03:01:35
```
Tracer from akka-typed is about ~2x slower, but has no dependencies.